src/Cli/dotnet/Installer/Windows/MsiPackageCache.cs

```diff
@@ -88,41 +88,63 @@ internal class MsiPackageCache : InstallerBase
         }
 
         /// <summary>
-        /// Creates the root package cache directory if it does not exist and configures the directory's ACLs. ACLs are still configured
-        /// if the directory exists.
-        /// </summary>
-        private void CreateRootDirectory()
-        {
-            CreateSecureDirectory(PackageCacheRoot);
-        }
-
-        /// <summary>
-        /// Creates the specified directory and secures it by configuring access rules (ACLs). If the parent
-        /// of the directory does not exist, it recursively walks the path back to ensure each parent directory
-        /// is created with the proper ACLs and inheritance settings.
+        /// Creates the specified directory and secures it by configuring access rules (ACLs) that allow sub-directories
+        /// and files to inherit access control entries. 
         /// </summary>
         /// <param name="path">The path of the directory to create.</param>
-        private void CreateSecureDirectory(string path)
+        /// <param name="resetAccessRules">When <see langword="true"/>, all existing access rules are removed before
+        /// setting up new rules.
+        /// </param>
+        /// <param name="log">The underlying setup log to use.</param>
+        public static void CreateSecureDirectory(string path, bool resetAccessRules = false, ISetupLogger log = null)
         {
             if (!Directory.Exists(path))
             {
-                CreateSecureDirectory(Directory.GetParent(path).FullName);
+                // Create the directory with the desired access rights if it doesn't exist. We don't use
+                // Directory.CreateDirectory as that can cause problems with inherited ACEs from
+                // the parent folders as access rules become cumulative. The acccess rules for the directory should all have
+                // object and container inheritance set to ensure access rights are inherited when creating files.
+                DirectorySecurity ds = new();
+                ds.SetOwner(s_AdministratorsSid);
```

Comment:
It seems like line 108-113 and line 133-138 is the same, I would suggest making a separate private function such as `SetAccessRules()`?